### DISCUSSION TITLE
Query optimization for performance

### DISCUSSION THREAD

**Discussion by SheaBelsky** [MichalLytek/typegraphql-prisma#258] [Q&A]
I'm running into some performance issues with my Prisma/TypeGraphQL project, and have been going through documentation/reading in order to find more information about the subject. More specifically, I am running into connection limit errors in an Azure environment, which are being caused by nested queries that cascade into lots of simultaneous SQL queries.

A great article by Prisma itself discusses query optimization, and how nested queries work:

https://www.prisma.io/docs/guides/performance-and-optimization/query-optimization-performance

Consider this query from my project:

```graphql
query candidates {
  __typename
  id
  user {
    __typename
    id
    name
  }
  resume {
    __typename
    id
    url
  }
}
```

According to the above Prisma article, one SQL query would execute for all `candidates`, and then two SQL queries would execute for each candidate (one for `user` and one for `resume`). **Sidenote:** Would love to know if there's a way to debug exactly how many queries are actually getting created in my actual application so I know if I'm actually making progress on this front or not.

`findUnique` is mentioned as a solve for this issue, but it doesn't look like TypeGraphQL uses that when compiling resolvers. Short of writing brand new resolvers that directly call Prisma's Fluent API, what else can I do to optimize queries in a TypeGraphQL project?

**Comment by MichalLytek** [comment]
> Sidenote: Would love to know if there's a way to debug exactly how many queries are actually getting created in my actual application so I know if I'm actually making progress on this front or not.

Check prisma docs about debug logs

> findUnique is mentioned as a solve for this issue, but it doesn't look like TypeGraphQL uses that when compiling resolvers. Short of writing brand new resolvers that directly call Prisma's Fluent API, what else can I do to optimize queries in a TypeGraphQL project?

The generated field resolvers are using `findUnique` so it benefits from Prisma's Dataloder optimizations:
```ts
@TypeGraphQL.FieldResolver(_type => [Problem], {
    nullable: false
  })
  async problems(@TypeGraphQL.Root() creator: Creator, @TypeGraphQL.Ctx() ctx: any, @TypeGraphQL.Args() args: CreatorProblemsArgs): Promise<Problem[]> {
    return getPrismaFromContext(ctx).creator.findUnique({
      where: {
        id: creator.id,
      },
    }).problems(args);
  }
  ```
